magma/types/decoupled.py

```diff
@@ -0,0 +1,174 @@
+from typing import Union
+from magma.bit import Bit
+from magma.primitives.mux import mux
+from magma.t import Type, In, Out, Flip
+from magma.protocol_type import MagmaProtocol
+from magma.tuple import Product, ProductKind
+
+
+class ReadyValidException(Exception):
+    pass
+
+
+class ReadyValidKind(ProductKind):
+    def _check_T(cls, T):
+        if not issubclass(T, (Type, MagmaProtocol)):
+            raise TypeError(
+                f"{cls}[T] expected T to be a subclass of m.Type or"
+                f" m.MagmaProtocol not {T}"
+            )
+
+    def __getitem__(cls, T: Union[Type, MagmaProtocol]):
+        cls._check_T(T)
```

Comment:
we might also want to get the undirected form of `T`. If `T` is directed, we could either (a) error/warn and do nothing, or (b) warn and get the undirected form of `T`.